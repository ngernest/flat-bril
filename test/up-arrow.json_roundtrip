{
  "functions": [
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        },
        {
          "name": "arrows",
          "type": "int"
        },
        {
          "name": "repeats",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "n",
            "arrows",
            "repeats"
          ],
          "dest": "ans",
          "funcs": [
            "up_arrow"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "ans"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "num",
          "type": "int"
        },
        {
          "name": "arrows",
          "type": "int"
        },
        {
          "name": "repeats",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "num"
          ],
          "dest": "ans",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "i",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "label": "loopstart"
        },
        {
          "args": [
            "i",
            "repeats"
          ],
          "dest": "keepgoing",
          "funcs": [],
          "labels": [],
          "op": "lt",
          "type": "bool"
        },
        {
          "args": [
            "keepgoing"
          ],
          "funcs": [],
          "labels": [
            "loopbody",
            "endloop"
          ],
          "op": "br"
        },
        {
          "label": "loopbody"
        },
        {
          "args": [
            "arrows",
            "one"
          ],
          "dest": "base_case",
          "funcs": [],
          "labels": [],
          "op": "le",
          "type": "bool"
        },
        {
          "args": [
            "base_case"
          ],
          "funcs": [],
          "labels": [
            "base",
            "arrowsgreaterthan1"
          ],
          "op": "br"
        },
        {
          "label": "base"
        },
        {
          "args": [
            "ans",
            "num"
          ],
          "dest": "ans",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "finally"
          ],
          "op": "jmp"
        },
        {
          "label": "arrowsgreaterthan1"
        },
        {
          "args": [
            "arrows",
            "one"
          ],
          "dest": "new_arrows",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "num",
            "new_arrows",
            "ans"
          ],
          "dest": "ans",
          "funcs": [
            "up_arrow"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "label": "finally"
        },
        {
          "args": [
            "i",
            "one"
          ],
          "dest": "i",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "loopstart"
          ],
          "op": "jmp"
        },
        {
          "label": "endloop"
        },
        {
          "args": [
            "ans"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "up_arrow",
      "type": "int"
    }
  ]
}
