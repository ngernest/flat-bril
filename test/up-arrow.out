@main
	op: call 		dest: ans		type: int		args: ["n", "arrows", "repeats"]		funcs: ["up_arrow"]
	op: print		args: ["ans"]	
@up_arrow
	op: const		dest: one		type: int		value: 1	
	op: id   		dest: ans		type: int		args: ["num"]	
	op: const		dest: i 		type: int		value: 1	
.loopstart
	op: lt   		dest: keepgoing		type: bool		args: ["i", "repeats"]	
	op: br   		args: ["keepgoing"]		labels: ["loopbody", "endloop"]
.loopbody
	op: le   		dest: base_case		type: bool		args: ["arrows", "one"]	
	op: br   		args: ["base_case"]		labels: ["base", "arrowsgreaterthan1"]
.base
	op: mul  		dest: ans		type: int		args: ["ans", "num"]	
	op: jmp  		labels: ["finally"]
.arrowsgreaterthan1
	op: sub  		dest: new_arrows		type: int		args: ["arrows", "one"]	
	op: call 		dest: ans		type: int		args: ["num", "new_arrows", "ans"]		funcs: ["up_arrow"]
.finally
	op: add  		dest: i 		type: int		args: ["i", "one"]	
	op: jmp  		labels: ["loopstart"]
.endloop
	op: ret  		args: ["ans"]	
